lib/shared/s3-packages.js

```diff
@@ -0,0 +1,212 @@
+ *
+ * @example
+ * const bucketUrl = s3Packages.getBucketUrlFromReleaseType(release.RELEASE_TYPE.PRODUCTION);
+ *
+ * if (bucketUrl) {
+ *   console.log(bucketUrl);
+ * }
+ */
+exports.getBucketUrlFromReleaseType = (releaseType) => {
+  if (releaseType === release.RELEASE_TYPE.PRODUCTION) {
+    return exports.BUCKET_URL.PRODUCTION;
+  }
+
+  if (releaseType === release.RELEASE_TYPE.SNAPSHOT) {
+    return exports.BUCKET_URL.SNAPSHOT;
+  }
+
+  return null;
+};
+
+/**
+ * @summary Get all remote versions from an S3 bucket
+ * @function
+ * @private
+ *
+ * @description
+ * We memoize based on the assumption that the received latest version
+ * number will not increase while the application is running.
+ *
+ * @param {String} bucketUrl - s3 bucket url
+ * @fulfil {String[]} - remote versions
+ * @returns {Promise}
+ *
+ * @example
+ * s3Packages.getRemoteVersions(s3Packages.BUCKET_URL.PRODUCTION).then((versions) => {
+ *   _.each(versions, (version) => {
+ *     console.log(version);
+ *   });
+ * });
+ */
+exports.getRemoteVersions = _.memoize((bucketUrl) => {
+  if (_.isNil(bucketUrl)) {
+    return Bluebird.reject(new Error(`Invalid bucket url: ${bucketUrl}`));
+  }
+
+  /* eslint-disable lodash/prefer-lodash-method */
+
+  return request.getAsync(bucketUrl)
+
+  /* eslint-enable lodash/prefer-lodash-method */
+
+    .get('body')
+    .then(xml.parseStringAsync)
+    .get('ListBucketResult')
+    .then((bucketResult) => {
+      return _.get(bucketResult, [ 'Contents' ], []);
+    })
+    .reduce((accumulator, entry) => {
+      const [ name, version ] = _.split(_.first(entry.Key), '/');
+
+      if (name === S3_PACKAGE_NAME) {
+        if (_.isNil(accumulator[version])) {
+          accumulator[version] = 1;
+        } else {
+          accumulator[version] += 1;
+        }
+      }
+
+      return accumulator;
+    }, [])
+    .then((versions) => {
+      return _.keys(_.pickBy(versions, (occurrences) => {
+        return occurrences >= NUMBER_OF_PACKAGES;
+      }));
+    })
+    .catch({
+      code: 'ENOTFOUND'
+    }, {
+      code: 'ETIMEDOUT'
+    }, () => {
+      return [];
+    });
+});
+
+/**
+ * @summary Get the latest available version from an S3 bucket
+ * @function
+ * @public
+ *
+ * @param {String} releaseType - release type
+ * @param {Object} options - options
+ * @param {String} [options.range] - semver range
+ * @fulfil {String} - latest version
+ * @returns {Promise}
+ *
+ * @example
+ * s3Packages.getLatestVersion(release.RELEASE_TYPE.PRODUCTION, {
+ *   range: '>=2.0.0'
+ * }).then((latestVersion) => {
+ *   console.log(`The latest version is: ${latestVersion}`);
+ * });
+ */
+exports.getLatestVersion = (releaseType, options = {}) => {
+
+  // For manual testing purposes
+  const ETCHER_FAKE_S3_LATEST_VERSION = process.env.ETCHER_FAKE_S3_LATEST_VERSION;
+  if (release.getReleaseType(ETCHER_FAKE_S3_LATEST_VERSION) === releaseType) {
+    return Bluebird.resolve(ETCHER_FAKE_S3_LATEST_VERSION);
+  }
```

Comment:
Hmmm, if the user has gone to the effort of setting `ETCHER_FAKE_S3_LATEST_VERSION` but `release.getReleaseType(ETCHER_FAKE_S3_LATEST_VERSION) !== releaseType`, I think it'd make more sense for this function to return `false` (as we do on line 199), rather than continuing to drop through to checking the available S3 versions. Perhaps we could print a message to STDERR and/or the devtools console warning them that the release-type of `ETCHER_FAKE_S3_LATEST_VERSION` doesn't match the current release-type of Etcher.

Also, the docstring for `.getLatestVersion` should be updated to mention that it returns `false` when it can't get the latest version. What value does `.getLatestVersion` return if (hypothetically) `.getRemoteVersions` returns an empty list?